test/Npgsql.Tests/MultipleHostsTests.cs

```diff
@@ -963,6 +963,117 @@ public void DataSource_with_single_host_is_not_supported()
             .With.Message.EqualTo(NpgsqlStrings.MultipleHostsMustBeSpecified));
     }
 
+    [Test, IssueLink("https://github.com/npgsql/npgsql/issues/4181")]
+    [Explicit("Fails until #4181 is fixed.")]
+    public async Task LoadBalancing_is_fair_if_first_host_is_down([Values]TargetSessionAttributes targetSessionAttributes)
+    {
+        using var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+        var endpoint = new IPEndPoint(IPAddress.Loopback, 0);
+        socket.Bind(endpoint);
+        var localEndPoint = (IPEndPoint)socket.LocalEndPoint!;
+        // Note that we Bind (to reserve the port), but do not Listen - connection attempts will fail.
+
+        await using var pRw1 = PgPostmasterMock.Start(state: Primary);
+        await using var pRw2 = PgPostmasterMock.Start(state: Primary);
+        await using var pR1 = PgPostmasterMock.Start(state: PrimaryReadOnly);
+        await using var pR2 = PgPostmasterMock.Start(state: PrimaryReadOnly);
+        await using var s1 = PgPostmasterMock.Start(state: Standby);
+        await using var s2 = PgPostmasterMock.Start(state: Standby);
+
+        var connectionString = new NpgsqlConnectionStringBuilder
+        {
+            Host = $"{localEndPoint.Address}:{localEndPoint.Port}," +
+                   $"{pRw1.Host}:{pRw1.Port}," +
+                   $"{pR1.Host}:{pR1.Port}," +
+                   $"{s1.Host}:{s1.Port}," +
+                   $"{pRw2.Host}:{pRw2.Port}," +
+                   $"{pR2.Host}:{pR2.Port}," +
+                   $"{s2.Host}:{s2.Port}",
+            ServerCompatibilityMode = ServerCompatibilityMode.NoTypeLoading,
+            LoadBalanceHosts = true,
+            TargetSessionAttributesParsed = targetSessionAttributes
+        }.ConnectionString;
+
+        var connections = Enumerable.Repeat(await OpenConnectionAsync(connectionString), 12).ToArray();
```

Comment:
I might be mistaken, but doesn't it just copy the exact same open connection 12 times?

Reply:
Yeah, you're right. I tried to be too smart here.